package lists (#249)
conan-io/conan-io.github.io: _posts/2023-06-28-Conan-bulk-package-operations.markdown

---
layout: post
comments: false
title: "Conan 2.0 launches bulk upload, download and remove operations with 'packages lists'"
description: "Learn how Conan 2.0 packages lists can be used to do bulk upload, download and remove operations, very useful in Continuous Integration flows"
---

In Conan 1.X, uploading the packages that were built from source in the last ``conan install --build=missing`` command was not a straightforward task.

> It was necessary to have some script that would get the output of the command, and then execute ``conan upload <ref>``, one reference at a time.

Reviewer: You could always just `conan upload *` if you started with an empty cache, you did not need any fancy scripts if the pipeline wasn't contaminated with other builds

Conan 2.0 superior design and architecture is allowing the quick implementation of better interfaces and UX tools and flows. We are happy to launch the "package lists" feature, that makes that functionality built-in.

> Conan 2.0.7 introduces a new json file format that is able to list a set of Conan recipes and packages, the "package list". This file can be used to perform bulk remove, download and upload operations.

Author: Yes, this is blog post, not technical documentation. The purpose is to convey an idea and to create excitement about a new feature, not to detail every technical aspect of the feature.

Let's start with a simple use case to define the concepts:

## Listing and downloading packages

We can do any ``conan list``, for example, to list all ``zlib`` versions above ``1.2.11``, the latest recipe revision,

> all Windows binaries for that latest recipe revision, and finally the latest package revision for every binary.

Author: I don't understand the past tense ``it was already possible``. I am describing that if you run that command you get a "pkg-list" with that.

Note that if we want to actually download something later, it is necessary to specify the ``latest`` package revision,
otherwise only the recipes will be downloaded.

```bash
$ conan list "zlib/[>1.2.11]#latest:*#latest" -p os=Windows --format=json -r=conancenter > pkglist.json
```

The output of the command is sent in ``json`` format to the file ``pkglist.json`` that looks like (simplified ``pkglist.json`` file):

```json
"conancenter": {
    "zlib/1.2.12": {
        "revisions": {
            "b1fd071d8a2234a488b3ff74a3526f81": {
                "timestamp": 1667396813.987,
                "packages": {
                    "ae9eaf478e918e6470fe64a4d8d4d9552b0b3606": {
                        "revisions": {
                            "19808a47de859c2408ffcf8e5df1fdaf": {
                            }
                        },
                        "info": {
                            "settings": {
                                "arch": "x86_64",
                                "os": "Windows"
                            }
                        }
                    }
                }
            }
        },
    "zlib/1.2.13": {
    }
  }
}
```

> The first level in the ``pkglist.json`` is the "origin" remote or "Local Cache" if the list happens in the cache. 

Author: There are explicit examples in this same blog post some lines below that uses ``downloaded.json`` filename.

In this case, as we listed the packages in ``conancenter`` remote, that will be the origin.

We can now do a download of these recipes and binaries with a single ``conan download`` invocation:

```bash

> $ conan download --list=pkglist.json -r=conancenter

Author: Of course, that is intrinsic to the ``conan download`` which already existed in Conan 1.X, there is nothing really new here.

security, reproducibility and robustness are very important would be building the packages from source, from a conan-center-index fork)

<p class="centered">
    <img  src="{{ site.baseurl }}/assets/post_images/2023-06-28/promote_air_gap.png" style="display: block; margin-left: auto; margin-right: auto;" alt="Promote packages in air-gapped environment with download+upload"/>
</p>

With the ``conan download`` command, we can generate a new package list containing the packages downloaded to the local cache:

```bash
$ conan download --list=pkglist.json -r=conancenter --format=json > downloaded.json
# Download the recipes and binaries in pkglist.json
# And stores the result in "downloaded.json"
```
It really doesn't matter if the input to ``conan download`` is already a list or a pattern. The resulting ``downloaded.json`` 
will be almost the same as the ``pkglist.json`` file, but in this case, the "origin" of those packages is the ``"Local Cache"`` 
(as the downloaded packages will be in the cache):

```json
"Local Cache": {
        "zlib/1.2.12": {
            "revisions": {
            }
        }
    }
```

> That means that we can now upload this same set of recipes and binaries to a different remote:

Author: It will not upload anything, it is not possible to upload packages that are not in the Local Cache. The only relevant origin for uploads is "Local Cache"

> That means that we can now upload this same set of recipes and binaries to a different remote:

Reviewer: Is that documented anywhere? Is the opposite true as well, you can not download from the local cache?

```bash
$ conan upload --list=downloaded.json -r=myremote -c
# Upload those artifacts to the same remote
```

One of the important points is that this works irrespective of the cache contents. It doesn't matter if the local cache
contains other artifacts from other commands. The packages uploaded to the server will be only the ones listed in the ``downloaded.json``.

> Note that this would be a **slow** mechanism to run promotions between different server repositories. Servers like

Author: I don't want to make it a section title. This is a note, that it might be slow, but in many cases it is the only possible approach, so not really a drawback, it is what it is.

> Artifactory provide ways to directly copy packages from one repository to another without using a client, 

Author: I'll remove the Artifactory reference.

that are orders of magnitude faster because of file deduplication, so that would be the recommended approach if the 2
repositories can directly talk to each other.

## Building from source and uploading packages

One of the most interesting flows is the one when some packages are being built in the local cache, with a 

> ``conan create`` or ``conan install --build=xxx`` command. Typically, we would like to upload the locally built

Author: ``conan create`` by default includes ``--build=currentpkg``, not sure what is the question.

packages to the server, so they don't have to be re-built again by others. But we might want to upload only
the built binaries, but not all others transitive dependencies, or other packages that we had previously in

> our local cache.

Author: > In cloud environments, where data transfers get pricey, uploading only the exact list, rather then all the transitive dependencies, offers a lot of potential.

It is not about cost, it is about correctness, control, etc.

<p class="centered">
    <img  src="{{ site.baseurl }}/assets/post_images/2023-06-28/build_and_upload.png" style="display: block; margin-left: auto; margin-right: auto;" alt="Build from source and upload built packages"/>
</p>

> It is possible to compute a package list from the output of a ``conan install``, ``conan create`` and ``conan graph info``

Reviewer: ```suggestion
It is possible to use the graph from the any of the commands ``conan install``, ``conan create`` and ``conan graph info`` to compute a package list which can then be used with upload
```

Author: > use the graph from the any of the commands

"from the any" of the commands doesn't sound correct english?

Reviewer: Good catch, that's a typo

> commands. Then, that package list can be used for the upload. Step by step:

Reviewer: ```suggestion
```

First let's say that we have our own package ``mypkg/0.1`` and we create it:

```bash
$ conan new cmake_lib -d name=mypkg -d version=0.1
$ conan create . --format=json > create.json
```

> This will create a json representation of the graph, with information of what packages have been built (that is, they contain ``"binary": "Build"``),

Author: Cache, Download, Update...

Not really relevant, this is not technical documentation, just some examples to communicate the concept of what can be done.

the ``create.json`` will look like (simplfied):

```json
    {
    "graph": {
        "nodes": {
            "0": {
                "ref": "conanfile",
                "id": "0",
                "recipe": "Cli",
                "context": "host",
                "test": false
            },
            "1": {
                "ref": "mypkg/0.1#f57cc9a1824f47af2f52df0dbdd440f6",
                "id": "1",
                "recipe": "Cache",
                "package_id": "2401fa1d188d289bb25c37cfa3317e13e377a351",
                "prev": "75f44d989175c05bc4be2399edc63091",
                "build_id": null,
                "binary": "Build"
            }
        }
      }
    }
```

We can compute a package list from this file, and then upload those artifacts to the server with:

```bash

> $ conan list --graph=create.json --graph-binaries=build --format=json > pkglist.json

Reviewer: Would it be possible to do this with a lockfile?

Author: No. 
You can for sure apply a lockfile when computing the graph with ``conan install``

# Create a pkglist.json with the known list of recipes and binaries built from sources

> $ conan upload --list=pkglist.json -r=myremote -c

Reviewer: Is upload the only command that does anything with a list or is this supported else where?

Author: What do you mean? The previous example contained a ``download --list`` command.

Author: Added other commands explanation in the intro

```

## Conclusions

The better Conan 2.0 architecture and design have allowed to implement "packages lists". This feature is a new powerful addition to the Conan CLI, allowing to do bulk operations over sets of packages that
previously required user custom automation (script files or Continuous Integration jobs), and now are conveniently provided as built-in. 

> Feedback is very welcome, please create a [Github issue for any question, comment or suggestion](https://github.com/conan-io/conan) about it.

Author: I prefer to link the https://github.com/conan-io/conan/issues one,  not the "new-choose" one, I'll change it.

---
layout: post
comments: false
title: "Conan 2.0 launches bulk upload, download and remove operations with 'packages lists'"

> meta_title: "Conan 2.0 packages-lists can be used to do bulk operations over lists of packages like upload, download or remove - Conan Blog"
> description: "Learn how Conan 2.0 packages lists can be used to do bulk upload, download and remove operations, very useful in Continuous Integration flows"
> keywords: "conan, conan2, packages lists, continuous integration, bulk operations"

Reviewer: Less is more :) You copied my last attempt but these did not have any effect I could find

```suggestion
description: "Learn how Conan 2.0 packages lists can be used to do bulk upload, download and remove operations, very useful in Continuous Integration flows"
```

---

> In Conan 1.X, when users wanted to upload the packages that were built from source in the last ``conan install --build=missing`` command, it was not a straightforward task.

Reviewer 2: ```suggestion
In Conan 1.X, uploading the packages that were built from source in the last ``conan install --build=missing`` command was not a straightforward task.
```

> It was necessary to have some script that will get the output of the command, and then execute ``conan upload <ref>``, one reference at a time.

Reviewer 2: ```suggestion
It was necessary to have some script that would get the output of the command, and then execute ``conan upload <ref>``, one reference at a time.
```

Conan 2.0 superior design and architecture is allowing the quick implementation of better interfaces and UX tools and flows. We are happy to launch the "package lists" feature, that makes that functionality built-in.

Conan 2.0.7 introduces a new json file format that is able to list a set of Conan recipes and packages, the "package list". This file can be used to perform bulk remove, download and upload operations.

Let's start with a simple use case to define the concepts:

## Listing and downloading packages

We can do any ``conan list``, for example, to list all ``zlib`` versions above ``1.2.11``, the latest recipe revision,
all Windows binaries for that latest recipe revision, and finally the latest package revision for every binary.
Note that if we want to actually download something later, it is necessary to specify the ``latest`` package revision,
otherwise only the recipes will be downloaded.

```bash

> $ conan list "zlib/[>1.2.11]#latest:*#latest" -p os=Windows --format=json -r=conancenter > pkglist.json

Reviewer 2: Maybe circling back to a link where the query syntax is explained in the docs would be good to have?

```

The output of the command is sent in ``json`` format to the file ``pkglist.json`` that looks like (simplified ``pkglist.json`` file):

```json
"conancenter": {
    "zlib/1.2.12": {
        "revisions": {
            "b1fd071d8a2234a488b3ff74a3526f81": {
                "timestamp": 1667396813.987,
                "packages": {
                    "ae9eaf478e918e6470fe64a4d8d4d9552b0b3606": {
                        "revisions": {
                            "19808a47de859c2408ffcf8e5df1fdaf": {
                            }
                        },
                        "info": {
                            "settings": {
                                "arch": "x86_64",
                                "os": "Windows"
                            }
                        }
                    }
                }
            }
        },
    "zlib/1.2.13": {
    }
  }
}
```

The first level in the ``pkglist.json`` is the "origin" remote or "Local Cache" if the list happens in the cache. 

> In this case, as we listed the packages in ``conancenter`` remote, that will be the origin.

Author: You can get recipe from one remote and package binary from another remote

repositories can directly talk to each other.

## Building from source and uploading packages

One of the most interesting flows is the one when some packages are being built in the local cache, with a 
``conan create`` or ``conan install --build=xxx`` command. Typically, we would like to upload the locally built
packages to the server, so they don't have to be re-built again by others. But we might want to upload only
the built binaries, but not all others transitive dependencies, or other packages that we had previously in
our local cache.

<p class="centered">
    <img  src="{{ site.baseurl }}/assets/post_images/2023-06-28/build_and_upload.png" style="display: block; margin-left: auto; margin-right: auto;" alt="Build from source and upload built packages"/>
</p>

It is possible to compute a package list from the output of a ``conan install``, ``conan create`` and ``conan graph info``
commands. Then, that package list can be used for the upload. Step by step:

First let's say that we have our own package ``mypkg/0.1`` and we create it:

```bash
$ conan new cmake_lib -d name=mypkg -d version=0.1
$ conan create . --format=json > create.json
```

> This will create a json representation of the graph, with information of what packages have been built ``"binary": "Build"``,

Reviewer 2: ```suggestion
This will create a json representation of the graph, with information of what packages have been built (that is, they contain ``"binary": "Build"``),
```

the ``create.json`` will look like (simplfied):

```json
    {
    "graph": {
        "nodes": {
            "0": {
                "ref": "conanfile",
                "id": "0",
                "recipe": "Cli",
                "context": "host",
                "test": false
            },
            "1": {
                "ref": "mypkg/0.1#f57cc9a1824f47af2f52df0dbdd440f6",
                "id": "1",
                "recipe": "Cache",
                "package_id": "2401fa1d188d289bb25c37cfa3317e13e377a351",
                "prev": "75f44d989175c05bc4be2399edc63091",
                "build_id": null,
                "binary": "Build"
            }
        }
      }
    }
```

We can compute a package list from this file, and then upload those artifacts to the server with:

```bash
$ conan list --graph=create.json --graph-binaries=build --format=json > pkglist.json
# Create a pkglist.json with the known list of recipes and binaries built from sources
$ conan upload --list=pkglist.json -r=myremote -c
```

## Conclusions

> The Conan 2.0 better architecture and design has allowed to implement "packages lists". This feature is a new powerful addition to the Conan CLI, allowing to do bulk operations over sets of packages that

Reviewer 2: ```suggestion
The better Conan 2.0 architecture and design have allowed to implement "packages lists". This feature is a new powerful addition to the Conan CLI, allowing to do bulk operations over sets of packages that
```

---
layout: post
comments: false
title: "Conan 2.0 launches bulk upload, download and remove operations with 'packages lists'"
description: "Learn how Conan 2.0 packages lists can be used to do bulk upload, download and remove operations, very useful in Continuous Integration flows"
---

In Conan 1.X, uploading the exact set of packages that were built from source in the last ``conan install --build=missing`` command was not a straightforward task.
It was necessary to have some script that would get the output of the command, and then iterate the resul, executing ``conan upload <ref>`` one reference at a time.

> We are happy to launch in Conan 2.0.7 the "packages lists" feature, that makes this highly requested functionality built-in.

Reviewer 2: ```suggestion
We are happy to launch the "packages lists" feature in Conan 2.0.7, which makes this highly requested functionality built-in.
```

This file can be generated as the ``--format=json`` of several commands. For example a ``conan list``
command like the following:

```bash
# The output file name is user defined
$ conan list "*:*" --format=json > pkglist.json
```

will generate a ``pkglists.json`` file that looks like this (simplified, not complete):

```json
{
    "Local Cache": {
        "zlib/1.2.12": {
            "revisions": {
                "b1fd071d8a2234a488b3ff74a3526f81": {
                    "timestamp": 1667396813.987,
                    "packages": {
                        "ae9eaf478e918e6470fe64a4d8d4d9552b0b3606": {
                            "revisions": {
                                "19808a47de859c2408ffcf8e5df1fdaf": {}
                            },
                            "info": {
                                "settings": {
                                    "arch": "x86_64",
                                    "os": "Windows"
                                }
                            }
                        }
                    }
                }
            }
        },
        "zlib/1.2.13": {
        }
    }

> }

Reviewer 3: May we add some `...` so it's clear that some of the parts of the file are missing?

Author: Problem is the json render doesn't understand that and prints it ugly?. For some reason it doesn't render fine here, could someone test the render, please?

```

The first level in the ``pkglist.json`` is the "location", it can be a remote or the local cache. In this case, we were listing
packages in the cache, so the origin is "Local Cache".

> The ``conan remove``, ``conan upload`` and ``conan download`` can also generate "packages lists" with the ``--format=json``

Reviewer 2: ```suggestion
The ``conan remove``, ``conan upload`` and ``conan download`` commands can also generate "packages lists" with the ``--format=json``
```

> formatter. For example ``conan upload ... -r=myremote --format=json`` will generate a packages lists for the "myremote" location

Reviewer 2: ```suggestion
formatter. For example ``conan upload ... -r=myremote --format=json`` will generate a packages list for the "myremote" location
```

as the first level in the resulting json file.

> Furthermore, it is also possible to generate a packages lists from the result of a ``conan create``, ``conan install`` and ``conan graph info``

Reviewer 2: ```suggestion
Furthermore, it is also possible to generate a packages list from the result of a ``conan create``, ``conan install`` and ``conan graph info``
```
